File: verifier/src/middleware/rateLimit.ts
/**
 * API rate limiting for Fastify using sliding window counters.
 * Uses Redis when available, falls back to in-memory per-process.
 */
import type { FastifyInstance, FastifyReply, FastifyRequest } from "fastify";

interface RateLimitOpts {
  requestsPerMinute: number;
  redis?: { zadd: Function; zremrangebyscore: Function; zcard: Function; expire: Function } | null;
}

const SKIP_PATHS = new Set(["/health"]);

export function registerRateLimit(app: FastifyInstance, opts: RateLimitOpts): void {
  const { requestsPerMinute, redis } = opts;
  const localWindows = new Map<string, number[]>();

  app.addHook("onRequest", async (request: FastifyRequest, reply: FastifyReply) => {
    if (SKIP_PATHS.has(request.url)) return;

    const clientIp = request.ip;
    const key = `ratelimit:${clientIp}`;
    const now = Date.now() / 1000;
    const windowStart = now - 60;

    let allowed: boolean;

    if (redis) {
      try {
        await redis.zremrangebyscore(key, 0, windowStart);
        const count = await redis.zcard(key) as number;
        if (count >= requestsPerMinute) {
          allowed = false;
        } else {
          await redis.zadd(key, now, String(now));
          await redis.expire(key, 120);
          allowed = true;
        }
      } catch {
        allowed = checkLocal(localWindows, key, now, windowStart, requestsPerMinute);
      }
    } else {
      allowed = checkLocal(localWindows, key, now, windowStart, requestsPerMinute);
    }

    if (!allowed) {
      reply.code(429).header("Retry-After", "60").send({
        error: "rate_limit_exceeded",
        retry_after_seconds: 60,
      });
    }
  });
}

function checkLocal(
  windows: Map<string, number[]>,
  key: string,
  now: number,
  windowStart: number,
  limit: number,
): boolean {
  const existing = windows.get(key) ?? [];
  const filtered = existing.filter((t) => t > windowStart);
  if (filtered.length >= limit) {
    windows.set(key, filtered);
    return false;
  }
  filtered.push(now);
  windows.set(key, filtered);
  return true;
}

File: fetcher/app/middleware/rate_limit.py
"""API-level rate limiting middleware for FastAPI.

Uses an in-memory sliding window counter per client IP. When Redis is available,
state is shared across workers via Redis; otherwise falls back to per-process
in-memory tracking (still useful for single-worker deploys).
"""
from __future__ import annotations

import time
from collections import defaultdict
from typing import Any

import structlog
from fastapi import Request, Response
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint

log = structlog.get_logger(__name__)


class RateLimitMiddleware(BaseHTTPMiddleware):
    def __init__(
        self,
        app: Any,
        *,
        requests_per_minute: int = 60,
        redis_client: Any = None,
    ):
        super().__init__(app)
        self._rpm = requests_per_minute
        self._redis = redis_client
        self._local: dict[str, list[float]] = defaultdict(list)

    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Response:
        if request.url.path in ("/health", "/docs", "/openapi.json"):
            return await call_next(request)

        client_ip = request.client.host if request.client else "unknown"
        key = f"ratelimit:{client_ip}"
        now = time.time()
        window_start = now - 60

        if self._redis:
            allowed = await self._check_redis(key, now, window_start)
        else:
            allowed = self._check_local(key, now, window_start)

        if not allowed:
            return Response(
                content='{"error":"rate_limit_exceeded","retry_after_seconds":60}',
                status_code=429,
                media_type="application/json",
                headers={"Retry-After": "60"},
            )

        return await call_next(request)

    async def _check_redis(self, key: str, now: float, window_start: float) -> bool:
        try:
            pipe = self._redis.pipeline()
            pipe.zremrangebyscore(key, 0, window_start)
            pipe.zcard(key)
            pipe.zadd(key, {str(now): now})
            pipe.expire(key, 120)
            results = await pipe.execute()
            count = results[1]
            return count < self._rpm
        except Exception:
            return self._check_local(key, now, window_start)

    def _check_local(self, key: str, now: float, window_start: float) -> bool:
        timestamps = self._local[key]
        self._local[key] = [t for t in timestamps if t > window_start]
        if len(self._local[key]) >= self._rpm:
            return False
        self._local[key].append(now)
        return True
